python-threatexchange/threatexchange/signal_type/index2.py

```diff
@@ -0,0 +1,166 @@
+            threshold (int): The maximum distance threshold for matches.
+
+        Returns:
+
+        """
+        query_array: np.ndarray = self._convert_pdq_strings_to_ndarray(queries)
+        limits, distances, indices = self.faiss_index.range_search(
+            query_array, threshold + 1
+        )
+
+        results: t.List[t.List[t.Any]] = []
+        for i in range(len(queries)):
+            matches = [idx.item() for idx in indices[limits[i] : limits[i + 1]]]
+            dists = [dist for dist in distances[limits[i] : limits[i + 1]]]
+            results.append(list(zip(matches, dists)))
+        return results
+
+    def __getstate__(self):
+        data = faiss.serialize_index(self.faiss_index)
+        return data
+
+    def __setstate__(self, data):
+        self.faiss_index = faiss.deserialize_index(data)
+
+    def _convert_pdq_strings_to_ndarray(
+        self, pdq_strings: t.Sequence[str]
+    ) -> np.ndarray:
+        """
+        Convert multiple PDQ hash strings to a numpy array.
+
+        Args:
+            pdq_strings (Sequence[str]): A sequence of 64-character hexadecimal PDQ hash strings
+
+        Returns:
+            np.ndarray: A 2D array of shape (n_queries, 256) where each row is the full PDQ hash as a bit array
+        """
+        hash_arrays = []
+        for pdq_str in pdq_strings:
+            try:
+                # Convert hex string to integer
+                hash_int = int(pdq_str, 16)
+                # Convert to binary string, padding to ensure 256 bits
+                binary_str = format(hash_int, "0256b")
+                # Convert to numpy array
+                hash_array = np.array(
+                    [int(bit) for bit in binary_str], dtype=np.float32
+                )
+                hash_arrays.append(hash_array)
+            except (ValueError, TypeError) as e:
+                raise ValueError(f"Invalid PDQ hash string: {pdq_str}") from e
+
+        # Convert list of arrays to a single 2D array
+        return np.array(hash_arrays, dtype=np.float32)
+
+
+Self = t.TypeVar("Self", bound="SignalTypeIndex2")
+
+PDQIndexMatch = IndexMatchUntyped[SignalSimilarityInfoWithIntDistance, IndexT]
+
+
+class SignalTypeIndex2(t.Generic[T]):
+    def __init__(
+        self,
+        threshold: int = DEFAULT_MATCH_DIST,
+        faiss_index: t.Optional[faiss.Index] = None,
+    ) -> None:
+        """
+        Initialize the PDQ index.
+
+        Args:
+            threshold (int): The maximum distance threshold for matches.
+            faiss_index (faiss.Index): An optional pre-existing FAISS index to use.
+        """
+        super().__init__()
+        if faiss_index is None:
+            # Use a simple brute-force FAISS index by default
+            faiss_index = faiss.IndexFlatL2(DIMENSIONALITY)
+        self.faiss_index = _PDQHashIndex(faiss_index)
+        self.threshold = threshold
+        self._deduper: t.Set[str] = set()
```

Comment:
From the meeting:

Should store this as a mapping from hash => faiss_id 

add(h, payload1)
add(h, playload2)

deduper: dict[hash, faiss_id] = dict[]
entries: list[int, payload] = list[]

existing_id = deduper.get(new_hash)
if existing_id is not None: 
   # Don't add to faiss!
   # we add to our internal entry mapping
   entries[existing_id].append(payload2)
else:
  # faiss id is 0 -> size
  next_id = len(deduper)
  faiss.add(h)
  entries.append([payload])
  deduper[h] = next_id

///
lookup(h) -> payloads
  faiss_id = faiss.search(h) 
  entries[faiss_id]

query(h) -> [payload1, payload2]
```